packages/x-data-grid/src/components/cell/GridCell.tsx

```diff
@@ -314,8 +313,8 @@ const GridCell = forwardRef<HTMLDivElement, GridCellProps>(function GridCell(pro
     [apiRef, field, rowId],
   );
 
-  const isCellRowSpanned = hiddenCells[rowId]?.[field] ?? false;
-  const rowSpan = spannedCells[rowId]?.[field] ?? 1;
+  const isCellRowSpanned = hiddenCells[rowId]?.[colIndex] ?? false;
+  const rowSpan = spannedCells[rowId]?.[colIndex] ?? 1;
```

Comment:
I have avoided chaging the state's shape, but for `rowspan` it was hard to avoid. I want the virtualizer to be as unaware as possible for the grid's column definition model, so for that the virtualizer needs to work with column indexes rather than `colDef.field`.